vulcan/lib/client/jsx/components/workflow/user_interactions/inputs/data_transformation.tsx

```diff
@@ -0,0 +1,236 @@
+import {WithInputParams, DataEnvelope} from './input_types';
+import {useSetsDefault} from './useSetsDefault';
+import {some, Maybe, withDefault, isSome} from '../../../../selectors/maybe';
+import {useMemoized} from '../../../../selectors/workflow_selectors';
+
+const useStyles = makeStyles((theme) => ({
+  dialog: {
+    maxWidth: '90vw',
+    maxHeight: '90vh'
+  },
+  subtitle: {display: 'inline'}
+}));
+
+function DataTransformationModal({
+  data,
+  onChange,
+  onClose
+}: {
+  data: any[][];
+  onChange: (data: Maybe<{[key: string]: any}>) => void;
+  onClose: () => void;
+}) {
+  const classes = useStyles();
+
+  const hotTableComponent = useRef<any>(null);
+  const hyperformulaInstance = HyperFormula.buildEmpty({
+    // to use an external HyperFormula instance,
+    // initialize it with the `'internal-use-in-handsontable'` license key
+    licenseKey: 'internal-use-in-handsontable'
+  });
+
+  return (
+    <>
+      <DialogTitle>
+        Transform your data (
+        <Typography className={classes.subtitle}>
+          <Link
+            target='_blank'
+            rel='noreferrer'
+            href='https://handsontable.github.io/hyperformula/guide/built-in-functions.html'
+          >
+            Supported functions
+          </Link>
+        </Typography>
+        )
+      </DialogTitle>
+      <DialogContent className={classes.dialog}>
+        <HotTable
+          ref={hotTableComponent}
+          settings={{
+            data: data,
+            colHeaders: true,
+            rowHeaders: true,
+            height: 'auto',
+            licenseKey: 'non-commercial-and-evaluation',
+            formulas: {
+              engine: hyperformulaInstance
+            },
+            contextMenu: {
+              items: {
+                col_left: {
+                  name: 'Insert column to left'
+                },
+                col_right: {
+                  name: 'Insert column to right'
+                },
+                remove_col: {},
+                undo: {},
+                redo: {},
+                clear_column: {}
+              }
+            }
+          }}
+        />
+      </DialogContent>
+      <DialogActions>
+        <Button
+          onClick={onClose}
+          startIcon={<CancelIcon />}
+          color='secondary'
+          variant='contained'
+        >
+          Cancel
+        </Button>
+        <Button
+          onClick={() => {
+            if (hotTableComponent.current) {
+              // hotTableComponent.current.hotInstance.getSourceData()
+              //   returns the raw formulas, unrendered, which we
+              //   set as input values to preserve any transformations.
+              const sourceData = hotTableComponent.current.hotInstance.getSourceData();
+              const data = hotTableComponent.current.hotInstance.getData();
+
+              onChange(
+                some({
+                  source_data: some(toJson(sourceData)),
```

Comment:
This output structure is now tightly coupled to the CWL outputs (in other words, your CWL steps have to include these two output names, exactly). I don't really like how that works, will revisit and see if there's a way to make it more generic...perhaps we can just return an array of inputs, and one is source and one is data. But that still requires the CWL author to know which output index is which set of data, which seems better but not great.

Reply:
I'm not quite following: as written, this results in whichever, singular output this ui output step names, to have an object, containing two keys, source_data and data.  the 'output' is still whatever name the CWL author would like, it just happens to be a nested object.

Reply:
Mmm, so the widget itself outputs a single Hash, yes. But then in [onChange](https://github.com/mountetna/monoetna/blob/cs/vulcan-data-transformation-widget/vulcan/lib/client/jsx/components/workflow/user_interactions/inputs/input_types.ts#L78-L84), it destructures this hash into multiple outputs.

It seemed like because this is a single UI widget controlling multiple outputs, I had to make changes to `input_types` to both `onChange` and `value`. The UI widget can pass additional information along to `onChange` (i.e. `multi=true` or some equivalent) to tell it to unpack or handle multiple output values. However, a UI widget cannot tell its parent components `UserInput` or `StepUserInput` that it requires multiple input values (`data` and `source_data` in this case)... thus, the packing and unpacking of multiple input / output values happens elsewhere, and this UI widget still only gets one input value and one output value. Thus it also unpacks the [input value](https://github.com/mountetna/monoetna/blob/cs/vulcan-data-transformation-widget/vulcan/lib/client/jsx/components/workflow/user_interactions/inputs/data_transformation.tsx#L211) and uses the part it needs.

Not sure if that makes sense. I can try to explain it better, if my thinking isn't clear...

Reply:
Ahh, so this brings us back to your original point about destructuring multiple outputs with either names or arrays.

What if you support both approaches: arrays OR named output matching?

IF the CWL author specifies an output by the same name as the multi output properties, those output names map to those output properties.

Otherwise,

each of the multiple output properties, in some fixed ordering (property ordering of the json object in onChange), is assigned to an output name corresponding to the array of output names in the CWL.

Any destructured output not referenced in any way (say the user only specifies 1 output but there are 3) are discarded with a warning in the system somehow.

I think the only thing that would concern me is if an output is matched to a name even when the CWL does not explicitly specify that name.  It would make analyzing the CWL for correctness much, much harder, and would make reading it tricky ("implicit output names" would make figuring out where values are coming from tricky).

Reply:
Hm, okay, let me try putting in more flexibility like that. I was trying another approach this morning, but I think it winds up as being more brittle than this current implementation, so going to trash it...

Reply:
And thank you for the suggestions! I've been a bit stuck trying to figure out how to make it more friendly, those ideas should definitely help.